internal/cmd/oohelperd/internal/nwcth/measure.go

```diff
@@ -0,0 +1,281 @@
+	h3Location   string
+}
+
+func Measure(ctx context.Context, creq *CtrlRequest) (*CtrlResponse, error) {
+	var cresp = &CtrlResponse{URLMeasurements: []*CtrlURLMeasurement{}}
+
+	redirected := make(map[string]bool, 100)
+
+	urlM, err := MeasureURL(ctx, creq, cresp, redirected)
+	if err != nil {
+		return nil, err
+	}
+	cresp.URLMeasurements = append(cresp.URLMeasurements, urlM.CtrlURLMeasurement)
+
+	n := 0
+	nextRequests := append(urlM.redirectedReqs, urlM.h3Reqs...)
+	for len(nextRequests) > n {
+		req := nextRequests[n]
+		n += 1
+		if len(redirected) == 20 {
+			// stop after 20 redirects
+			break
+		}
+		if _, ok := redirected[req.HTTPRequest]; ok {
+			continue
+		}
+		redirected[req.HTTPRequest] = true
+		urlM, err := MeasureURL(ctx, req, cresp, redirected)
+		if err != nil {
+			return nil, err
+		}
+		cresp.URLMeasurements = append(cresp.URLMeasurements, urlM.CtrlURLMeasurement)
+		nextRequests = append(nextRequests, urlM.redirectedReqs...)
+	}
+
+	return cresp, nil
+}
+
+// Measure performs the measurement described by the request and
+// returns the corresponding response or an error.
+func MeasureURL(ctx context.Context, creq *CtrlRequest, cresp *CtrlResponse, redirected map[string]bool) (*MeasureURLResult, error) {
+	// parse input for correctness
+	URL, err := url.Parse(creq.HTTPRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	// create URLMeasurement struct
+	urlMeasurement := &CtrlURLMeasurement{
+		URL:       URL.String(),
+		DNS:       nil,
+		Endpoints: []CtrlEndpointMeasurement{},
+	}
+
+	// dns: start
+	dns := DNSDo(ctx, &DNSConfig{
+		Domain: URL.Hostname(),
+	})
+
+	urlMeasurement.DNS = &dns
+
+	enpnts := getEndpoints(dns.Addrs, URL)
+	addrs := mergeEndpoints(enpnts, creq.TCPConnect)
+
+	if len(addrs) == 0 {
+		return nil, errors.New("no valid IP address to measure")
+	}
+
+	wg := new(sync.WaitGroup)
+	out := make(chan *MeasureEndpointResult, len(enpnts))
+	for _, endpoint := range enpnts {
+		wg.Add(1)
+		go MeasureEndpoint(ctx, creq, URL, endpoint, wg, out)
+	}
+	wg.Wait()
+	close(out)
+
+	h3Reqs := []*CtrlRequest{}
+	redirectedReqs := []*CtrlRequest{}
+	for m := range out {
+		urlMeasurement.Endpoints = append(urlMeasurement.Endpoints, m.CtrlEndpoint)
+		if m.httpRedirect != nil {
+			req := &CtrlRequest{HTTPCookieJar: m.httpRedirect.jar, HTTPRequest: m.httpRedirect.location, HTTPRequestHeaders: m.httpRedirect.httpRedirectReq.Header}
+			redirectedReqs = append(redirectedReqs, req)
+		}
+		if m.h3Location != "" {
+			req := &CtrlRequest{HTTPRequest: m.h3Location}
+			h3Reqs = append(h3Reqs, req)
+		}
+	}
+	return &MeasureURLResult{CtrlURLMeasurement: urlMeasurement, h3Reqs: h3Reqs, redirectedReqs: redirectedReqs}, nil
+}
+
+func MeasureEndpoint(ctx context.Context, creq *CtrlRequest, URL *url.URL, endpoint string, wg *sync.WaitGroup, out chan *MeasureEndpointResult) {
+	defer wg.Done()
+	endpointResult := &MeasureEndpointResult{}
+	measureFactory[URL.Scheme](ctx, creq, endpoint, wg, endpointResult)
```

Comment:
What would happen here in the unlikely event in which `URL.Scheme` is not part of the `measureFactory` map?

Reply:
I find very confusing this design where the return value is passed by pointer as the last argument. Would it be possible to change this `measureFactory` to return a `MeasureEndpointResult` instead?

Reply:
> What would happen here in the unlikely event in which URL.Scheme is not part of the measureFactory map?

`MeasureURL` throws an error when the URL Scheme (checkout `getEndpoints()`) is not either "http", "https", "h3", or "h3-29". So I thought it would be safe to assume that this case is covered. Is that sufficient?